BinaryTree.java

```diff
@@ -0,0 +1,171 @@
+import java.util.ArrayList;
+
+public class BinaryTree<T> {
+	Node root;
+
+	public BinaryTree(T root) {
+		this.root = new Node(root);
+	}
+
+	public boolean add(T current, T newNode, boolean addToLeft) {
+		Node currentNode;
+		if (current.equals(root.getVal())) {
+			currentNode = root;
+		} else {
+			currentNode = findNode(root.left, root.right, current);
+		}
+		if (current != null) {
+			Node nodeToAdd = new Node(newNode);
+			nodeToAdd.setParent(currentNode);
+
+			if (addToLeft) {
+				if (currentNode.getLeft() != null) {
+					return false;
+				}
+				currentNode.setLeft(nodeToAdd);
+			} else {
+				if (currentNode.getRight() != null) {
+					return false;
+				}
+				currentNode.setRight(nodeToAdd);
+			}
+			return true;
+		}
+		return false;
+	}
+
+	/**
+	 * Provides a string containing the ancestors of a specific node. If the
+	 * required node is not in the tree or it is the root node, the returned
+	 * string contains message explaining that the node doesn't have ancestors
+	 * in the tree.
+	 * 
+	 * @param val
+	 *            - The node for which the ancestors are required
+	 * @return - A string containing the ancestors of the node
+	 */
+	public String printAncestors(T val) {
+		ArrayList<Node> ancestors = findAncestors(val);
+		StringBuilder str = new StringBuilder();
+
+		if (ancestors.size() == 0) {
+			return "This node doesn't have ancestors in the tree";
+		}
+		for (Node a : ancestors) {
+			str.append(a.getVal().toString());
+			str.append(" ");
+		}
+
+		return str.toString();
+	}
+
+	/**
+	 * Provides a list that contains the ancestors of a specific node If the
+	 * node doesn't have ancestors in the tree, the returned list is empty.
+	 */
+	private ArrayList<Node> findAncestors(T val) {
+		Node node = findNode(root.getLeft(), root.getRight(), val);
+		ArrayList<Node> ancestors = new ArrayList<>();
+
+		while (node != null && node.getParent() != null) {
+			node = node.getParent();
+			ancestors.add(node);
+		}
+		return ancestors;
+	}
+
+	/**
+	 * Given the value of a node, it finds the specific node in the tree. If
+	 * there is no node for the requested value, the returned value is null
+	 */
+	private Node findNode(Node left, Node right, T val) {
```

Comment:
May be cleaner to simply pass in the root instead of left and right, then get left and right within the method.